Add Gornir constructor

nleiva opened the pull request:
This is work in progress. I got rid of `pkg/plugins/inventory` to avoid a circular dependency, therefore I moved `fromYAMLFile` to gornir package and made it an unexported function; users would need to use `WithInventory` instead.

- [x] Create Constructor/Build function
- [x] Update Examples
- [x] Add test cases
- [x] Update Readme file

nleiva commented:
Added a two basic test cases for builder. You can run it from the root of the repo with `go test -v -cover ./...`. `-cover` only to show the current coverage: `27.7% of statements`.

dbarrosop commented on pkg/gornir/gornir.go:
```diff
@@ -16,6 +19,89 @@ type Gornir struct {
 }
 
+// NewGornir is a Gornir constructor.
+func NewGornir() *Gornir {
```
I'd personally just skip this, it doesn't provide much value

nleiva replied:
Yeah, it doesn't provide any functionality right now. However is a placeholder to insert any defaults in the future.

dbarrosop commented on pkg/gornir/gornir.go:
```diff
@@ -16,6 +19,89 @@ type Gornir struct {
+type SetOption func(r *Gornir) error
+
+// Build is a Gornir constructor with options.
+func Build(opts ...SetOption) (*Gornir, error) {
```
I'd rename to `New` as it will read as `gornir.New()`

nleiva replied:
Yeah, makes sense.

dbarrosop commented:
That `SetOpt` pattern is quite interesting, I am taking notes :P However, not sure it works for what I had in mind. The main reason being that it doesn't let you build on the previous object. For instance, what I'd like to be able to do is this:

``` golang
gr := gornir.New(...)
grFiltered := gr.Filter(...) // gr is not modified so I can still use it
grSomeOtherFilter := gr.Filtered(...) // I still have access to gr and grFiltered
```
Same with `WithLogger`:

``` golang
gr := gornir.New(...) // I have a "default" logger
grFilteredWithDifferentLogger := gr.Filter(...).WithLogger(customLoggerOnlyForThis)
```

If you check how the `Logger` is used you will see that pattern being used to build the logger incrementally at different stages. For instance:

1. Instantiate a logger with a few fields and pass it: https://github.com/nornir-automation/gornir/blob/master/pkg/gornir/gornir.go#L54
2. Add more relevant fields without knowing/caring what it was done to it previously: https://github.com/nornir-automation/gornir/blob/master/pkg/plugins/runner/parallel.go#L22

> I moved fromYAMLFile to gornir package and made it an unexported function; users would need to use WithInventory instead.

I think this is due to a misunderstanding on what I had in mind (it's very early in the process and documentation is far from complete so I am sure we will have a few of those :) ). The `inventory` plugins are basically functions that generate an `*Inventory` from various sources and I don't think any of those should be in the main `gornir` package. You have seen already the `FromYAML` one but you can imagine in the future having there a `FromNSOT` or `FromNetbox`. With your change this doesn't seem plausible anymore.

The second issue is its new home, I think that addressing my previous comment will allow you to move it back to the plugins namespace but if it's still being an issue I'd suggest creating an interface for those. For instance:

``` golang
type InventoryPlugin interface {
   New() (*Inventory, error)
}
```

Then:

``` golang
type FromYAML struct {
   HostsFile string
}

func (f *FromYAML) New() (*Inventory, error) {
   // probably same code as now
}
```
However, I am not sure we need anything as sophisticated.

Does any of this make sense? Opinions? Thoughts?

nleiva commented:
> For instance, what I'd like to be able to do is this:
> 
> ```go
> gr := gornir.New(...)
> grFiltered := gr.Filter(...) // gr is not modified so I can still use it
> grSomeOtherFilter := gr.Filtered(...) // I still have access to gr and grFiltered
> ```
> 
> Same with `WithLogger`:
> 
> ```go
> gr := gornir.New(...) // I have a "default" logger
> grFilteredWithDifferentLogger := gr.Filter(...).WithLogger(customLoggerOnlyForThis)
> ```

I think I can make this work by making a Gornir the receiver of the function. Let me make the changes to see if it makes sense, it might not :-)

nleiva commented:
> > I moved fromYAMLFile to gornir package and made it an unexported function; users would need to use WithInventory instead.
> 
> I think this is due to a misunderstanding on what I had in mind (it's very early in the process and documentation is far from complete so I am sure we will have a few of those :) ). The `inventory` plugins are basically functions that generate an `*Inventory` from various sources and I don't think any of those should be in the main `gornir` package. You have seen already the `FromYAML` one but you can imagine in the future having there a `FromNSOT` or `FromNetbox`. With your change this doesn't seem plausible anymore.
> 

Yep. I hear you, `WithInventory(f string)` should become `WithInventory(r io.Reader)` i.e. something you can read from or similar. We will need to define this interface to decouple Gornir from the implementation details (`YAML`, `Netbox` or else), I'm with you 100%. I just didn't want to introduce that level of abstraction just yet, wanted to take one step at a time.

nleiva commented:
> The second issue is its new home, I think that addressing my previous comment about the `With` pattern will allow you to move it back to the plugins namespace but if it's still being an issue I'd suggest creating an interface for those. For instance:
> 
> ```go
> type InventoryPlugin interface {
>    New() (*Inventory, error)
> }
> ```
> 
> Then:
> 
> ```go
> type FromYAML struct {
>    HostsFile string
> }
> 
> func (f *FromYAML) New() (*Inventory, error) {
>    // probably same code as now
> }
> ```
> 
> However, I am not sure we need anything as sophisticated.
> 
> Does any of this make sense? Opinions? Thoughts?

Yeah, let me make some changes and see if this would fly, otherwise we can discard it, no problem.

dbarrosop commented:
> WithInventory(r io.Reader)

If we need to create such iinterface I'd suggest doing what I described before:

```
type InventoryPlugin interface {
   New() (*Inventory, error)
}
```

``` golang
type FromYAML struct {
   HostsFile string
}

func (f *FromYAML) New() (*Inventory, error) {
   // probably same code as now
}
```

```
WithInventory(FromYAML{HostsFile: "hosts.yaml"})
```

Then `WithInventory` would just call the `.New()` method. It's easy enough for now and should be flexible enough to allow to support anything.

nleiva commented:
> Then `WithInventory` would just call the `.New()` method. It's easy enough for now and should be flexible enough to allow to support anything.

Yeah, I'm just thinking `New` is for concrete types, not interfaces as they carry no value. But I think the idea is correct, this is how we would abstract away the source of the info to the Gornir package. Let me process this a bit and see what I can come up with, most likely what you just described.

nleiva commented:
Ok, made a few adjustments.

1. Created the `InventoryPlugin` interface. I went with `Create` instead of `New` at it better describes the behavior. An `InventoryPlugin` creates an `Inventory`. Also, notice I use value semantics (`Inventory`) over pointer semantics (`*Inventory`), which I think is better in this case (also in other parts of the code I'd like to change on another PR).

```go
type InventoryPlugin interface {
	Create() (Inventory, error)
}
```

You'd pass a plugin implementation to `WithInventory` option when building a Gornir. We have successfully then decoupled the implementation details of the plugin, so we can add others easily.

```go
file := "/go/src/github.com/nornir-automation/gornir/examples/hosts.yaml"
plugin := inventory.FromYAML{HostsFile: file}

gr, err := gornir.Build(
	gornir.WithInventory(plugin),
)
```

2. Added a Build function with a Gornir receiver. So you can now either create a Gornir from scratch

```go
original, err := gornir.Build(
				gornir.WithInventory(plugin),
				gornir.WithLogger(logger),
			)
```

Or build a new one (`filtered`) from an existing one (`original`):

```go
filtered, err := original.Build(gornir.WithFilter(filter))
```

3. Added more test cases, so code coverage is around 35%.

dbarrosop commented:
Awesome work :)

Give me a couple of days for a more thorough review though (local holiday here tomorrow)

dbarrosop commented on pkg/gornir/gornir.go:
```diff
@@ -16,6 +17,77 @@ type Gornir struct {
+func WithLogger(l Logger) SetOption {
+	return func(g *Gornir) error {
+		if l == nil {
+			errors.New("didn't receive a valid logger")
```
missing return

nleiva replied:
It returns a SetOption function: `return func(g *Gornir) error{...}`

dbarrosop commented on pkg/gornir/gornir.go:
```diff
@@ -16,6 +17,77 @@ type Gornir struct {
+func WithFilter(f FilterFunc) SetOption {
+	return func(g *Gornir) error {
+		if f == nil {
+			errors.New("didn't receive a valid filter function")
```
missing return

nleiva replied:
It returns a SetOption function: `return func(g *Gornir) error{...}`

nleiva replied:
> Awesome work :)
> 
> Give me a couple of days for a more thorough review though (local holiday here tomorrow)

No rush, this gives me time to re-think this.

dbarrosop commented:
I found some time to test very quickly the `Build` pattern you implemented and I am afraid it doesn't work. You can verify it by doing this:

In TestBuild` you can change:

```
if len(filtered.Inventory.Hosts) != tc.length {
```

for:

```
if len(filtered.Inventory.Hosts) == len(original.Inventory.Hosts) {
```

You can also add `fmt.Printf("%+v\n", original.Inventory.Hosts)` in the next line, you will notice that the filter also changed the original object.

I suggest following this pattern instead; https://github.com/tmrts/go-patterns/blob/master/creational/builder.md From a usability perspective it feels nicer (as you concatenate functions) and from an implementation perspective you just need to do the same that the original `Filter` method is doing.

nleiva commented:
Thanks for reviewing this. Let me check why it fails, might give it another try out of curiosity. Otherwise switch to the pattern you suggest in a different PR.

nleiva commented:
I believe I fixed the issue, it had to do with the use of pointer semantics. At some point I will probably propose some changes regarding this.
 
On the other hand, I will implement the other builder patter as well in another PR/branch so we can compare the two and pick one.

Added the following test case that first checks if the length of the filtered Gornir is as expected and then validates the original length is unmodified.

```go
original, err := gornir.Build(
	gornir.WithInventory(plugin),
	gornir.WithLogger(logger),
)
olen := len(original.Inventory.Hosts)
// ...
filtered, err := original.Build(gornir.WithFilter(tc.filter))
// ...
if len(filtered.Inventory.Hosts) != tc.length {
	t.Fatalf("Filtered Inventory Length in Test Case '%s' is %v, want %v",
		tc.name, len(filtered.Inventory.Hosts), tc.length)
}
if len(original.Inventory.Hosts) != olen {
	t.Fatalf("Oringinal Inventory Length in Test Case '%s' is %v, want %v",
		tc.name, len(original.Inventory.Hosts), olen)
}
```

nleiva commented:
> I suggest following this pattern instead; https://github.com/tmrts/go-patterns/blob/master/creational/builder.md From a usability perspective it feels nicer (as you concatenate functions) and from an implementation perspective you just need to do the same that the original `Filter` method was doing.

One thing that I want to point out about this patter is that it doesn't feel like Go to me (concatenating functions). Going back to your example:

```go
gr := gornir.New(...) // I have a "default" logger
grFilteredWithDifferentLogger := gr.Filter(...).WithLogger(customLoggerOnlyForThis)
```

This means that nothing can go wrong with `Filter(...)` or `WithLogger(customLoggerOnlyForThis)`, as they won't return an error, which we typically do in Go, like in `Create() (Inventory, error)`. This is why I prefer the pattern I used, because it will fail to create the object if any of the options have an error.

Anyways, I will be sketching the other patter to compare it.

nleiva commented:
Closing this one, we are most likely going with a different alternative.
